src/librustc/ty/sty.rs

```diff
@@ -2402,17 +2402,84 @@ pub struct Const<'tcx> {
 static_assert_size!(Const<'_>, 48);
 
 impl<'tcx> Const<'tcx> {
+    /// Literals and const generic parameters are eagerly converted to a constant, everything else
+    /// becomes `Unevaluated`.
+    pub fn from_hir_anon_const(tcx: TyCtxt<'tcx>, def_id: DefId, ty: Ty<'tcx>) -> &'tcx Self {
```

Comment:
Hmm, what if you remove the `ty` argument and get it from `tcx.type_of(def_id)`?

Also `from_anon_const` could be an okay name. I still would prefer if this handled all `Unevaluated` and only for  *local* `hir::Node::AnonConst` specifically it would bother to look at the body but I suppose that's inconsistent.

Oh, I know, you can change `DefId` to `LocalDefId`, then finally everything should be good.

Reply:
Not passing `Substs` is fine in retrospect, it makes this constructor like `tcx.type_of` but "value of", instead, I suppose (well, "unevaluated value of").